hasura/graphql-engine
# Implemented graceful shutdown for websockets

### Description
Implements graceful shutdown for websockets. When the sockets are not actively receiving data they will receive a closing message and the server will promptly shut down.

### Affected components 
- Server

### Solution and Design
The general design I followed:

* Keep the knowledge of how to close connections inside the Websocket
  module
* Add a field to `WsServer` to hold a channel, the channel signals when
  the application is stopping
* Return a callback to `Main` that can be invoked to send a message to
  the channel

### Steps to test and verify

* Start the server
* Use `gql` with any subscription
* Send a `TERM` signal to the process. It should stop immediately
